### PR TITLE
Avoid generating cover letters for single-patch Pull Requests

### PR DESCRIPTION
**Pull request by dscho**
This closes https://github.com/gitgitgadget/gitgitgadget/issues/85.

It also fulfills a wish by @gitster, at long last.

### CODE REVIEW

File: lib/patch-series.ts [hunk 1]
```diff
@@ -868,10 +907,10 @@ export class PatchSeries {
...
         const commitRange =
             `${this.project.baseCommit}..${this.project.branchName}`;
-        if (!this.coverLetter &&
-            1 < parseInt(await git(["rev-list", "--count", commitRange],
-                                   { workDir: this.project.workDir }),
-                         10)) {
+        const patchCount =
+            await revListCount(["--no-merges", commitRange],
+                               this.project.workDir);
+        if (!this.coverLetter && 1 < patchCount) {
```
Comment (webstech) [inline lib/patch-series.ts]: Should `this.patchCount` be used here instead of a local?

File: lib/patch-series.ts [hunk 2]
```diff
@@ -896,7 +935,7 @@ export class PatchSeries {
         if (subjectPrefix) {
             args.push("--subject-prefix=" + subjectPrefix);
         }
-        if (this.coverLetter) {
+        if (this.coverLetter && patchCount > 1) {
```
Comment (webstech) [inline lib/patch-series.ts]: It seems like the code around line ~913 could be moved here.  Might make the logic a little clearer.
```
if (this.patchCount > 1 ) {
   if (this.coverLetter) {
      args.push("--cover-letter");
   } else {
        throw new Error("Branch " + this.project.branchName
                + " needs a description");
  }
}
```

File: lib/git.ts [hunk 3]
```diff
@@ -128,6 +128,31 @@ export async function revParse(argument: string, workDir?: string) {
     return result.exitCode ? undefined : trimTrailingNewline(result.stdout);
 }
 
+/**
+ * Call `git rev-list --count` to count objects in a commit range.
+ *
+ * @param { string[] } rangeArgs the arguments to pass to `git rev-list`
+ * @param { string | undefined } workDir
+ *    the working directory in which to run `git rev-parse`
+ * @returns { number | undefined } the number of commits in the commit range
```
[the remark this answers is not in the document]
Comment (dscho) [reply lib/git.ts]: You're right! Will fix.

File: lib/git.ts [hunk 4]
```diff
@@ -128,6 +128,31 @@ export async function revParse(argument: string, workDir?: string) {
...
+ */
+export async function revListCount(rangeArgs: string | string[],
+                                   workDir?: string):
+    Promise<number> {
+    let gitArgs: string[];
+    if (typeof(rangeArgs) === "string") {
+        gitArgs = ["rev-list", "--count", rangeArgs];
+    } else {
+        gitArgs = ["rev-list", "--count", ...rangeArgs];
+    }
```
[the remark this answers is not in the document]
Comment (dscho) [reply lib/git.ts]: Sure!

File: lib/git.ts [hunk 5]
```diff
@@ -128,6 +128,31 @@ export async function revParse(argument: string, workDir?: string) {
...
+/**
+ * Call `git rev-list --count` to count objects in a commit range.
+ *
+ * @param { string[] } rangeArgs the arguments to pass to `git rev-list`
+ * @param { string | undefined } workDir
+ *    the working directory in which to run `git rev-parse`
+ * @returns { number | undefined } the number of commits in the commit range
+ */
+export async function revListCount(rangeArgs: string | string[],
+                                   workDir?: string):
```
[the remark this answers is not in the document]
Comment (dscho) [reply lib/git.ts]: Thanks!

You know, if you point out those kinds of issues only a few hundred more times, I _might_ learn to use default values after all... 😄

File: lib/patch-series.ts [hunk 6]
```diff
@@ -783,6 +805,23 @@ export class PatchSeries {
                                                  mails.length > 1, footers);
         }
 
+        if (mails.length === 1 && this.coverLetter) {
```
[the remark this answers is not in the document]
Comment (dscho) [reply lib/patch-series.ts]: I had this there originally, but then the "Submitted-As:" footers and the range-diff are inserted _before_ the cover letter, which is not what I want.

[the remark this answers is not in the document]
Comment (dscho) [reply lib/patch-series.ts]: I added a code comment to clarify why we have to do this at that point (and not earlier).

File: lib/patch-series.ts [hunk 7]
```diff
@@ -783,6 +805,23 @@ export class PatchSeries {
...
+            if (this.patchCount !== 1) {
+                throw new Error(`Patch count mismatch: ${mails.length} vs ${
+                    this.patchCount}`);
+            }
```
[the remark this answers is not in the document]
Comment (dscho) [reply lib/patch-series.ts]: I guess it's a bit overzealous, but I am really anxious not to break anything with this patch, so I'd like to keep this check.

File: lib/patch-series.ts [hunk 8]
```diff
@@ -783,6 +805,23 @@ export class PatchSeries {
...
+            // Need to insert it into the first mail
+            const splitAtTripleDash = mails[0].match(/([^]*?\n---\n)([^]*)$/);
+            if (!splitAtTripleDash) {
+                throw new Error(`No \`---\` found in\n${mails[0]}`);
+            }
+            console.log(`Insert cover letter into\n${mails[0]}\nwith match:`);
+            console.log(splitAtTripleDash);
+            mails[0] = splitAtTripleDash[1] +
+                this.coverLetter + "\n" + splitAtTripleDash[2];
```
[the remark this answers is not in the document]
Comment (dscho) [reply lib/patch-series.ts]: That would result in a line 93 columns long, that's why I avoided it.

File: lib/patch-series.ts (hunk 1 above)
Comment (dscho) [reply lib/patch-series.ts]: D'oh!

File: lib/patch-series.ts (hunk 2 above)
Comment (dscho) [reply lib/patch-series.ts]: Good idea! I did not find any good commit to squash this into, so I left it as an add-on patch.

File: lib/patch-series.ts (hunk 8 above)
[the remark this answers is not in the document]
Comment (webstech) [reply lib/patch-series.ts]: I wasn't suggesting it had to be on one line.  You showed me how useful template variables can be to split across lines and ignore white space.  Not a big deal - thought you leaned that way.  No idea of the cost of this way vs template strings, which are a kind of a sprintf.

**Comment by dscho** [comment]
> If I understand this correctly, the PR title/description is being placed after the '---' and before the change summary and the footers are being placed after the change summary and before the diff (or whatever else may show up there). I don't want to stop this going through will this be okay with the mailing list? I guess we will find out.

Yes, this is as intended, and this is how other single-patch contributions work, so I assume it is desired on the Git mailing list ;-)

> Since the default PR text is the commit message, we may want to add some filtering in the future to avoid being too redundant in the email.

That's a very valid concern... Let me see what I can do about this.

**Comment by dscho** [comment]
> > Since the default PR text is the commit message, we may want to add some filtering in the future to avoid being too redundant in the email.
> 
> That's a very valid concern... Let me see what I can do about this.

Hmm. This does not seem to be trivial, as I will need the commit message of the first commit, and I would love to avoid duplicating the effort in `checkCommits()`. But then, maybe it is not all that important to avoid one extra GitHub API call.

In any case, I plan on working on this later, as I am in the middle of the release engineering for Git for Windows v2.25.0-rc2 and I would like to send out a single patch right away, _without_ a separate cover letter.

So I am going to merge this here PR first.

**Pull request merged**